lib/pbench/agent/tool_data_sink.py

```diff
@@ -186,6 +190,408 @@ def terminate(self):
+        self.control_file = os.path.join(self.benchmark_run_dir, "pmie.d")
+        self.log_dir = os.path.join(self.benchmark_run_dir, "pmie")
+        self.temp_hosts = list(self.host_tools_dict.keys())
+        self.hosts = []
+
+        self.check_connection(self.temp_hosts)
+        self.build_pmie_config()
+        self.build_pmie_control_file(self.hosts)
+
+    def start_pmie(self):
+        """ invoke pmie and start using pmie_check -c """
+        # code to start logging
+        self.logger.info("starting pmie")
+        command = PCP.pmGetConfig("PCP_BINADM_DIR") + "/pmie_check -c {}".format(
+            self.control_file
+        )
+        self.logger.debug("start pmie command: %s", command)
+        out = subprocess.Popen(
+            [command], stdout=subprocess.PIPE, stderr=subprocess.STDOUT, shell=True
+        )
+        stdout, stderr = out.communicate()
+        self.logger.debug("start pmie command output:%s", stdout.decode())
+
+        if stdout.decode() != "":
+            # pmie failed. Log it
+            self.logger.error(
+                "not able to start pmie successfully due to:{}".format(stdout)
+            )
+        else:
+            self.logger.info("pmie started successfully")
+
+    def stop_pmie(self):
+        """ stop pmie using pmie_check -s """
+        # stopping pmie
+        self.logger.info("stopping pmie")
+        command = PCP.pmGetConfig("PCP_BINADM_DIR") + "/pmie_check -s -c {}".format(
+            self.control_file
+        )
+        self.logger.debug("stop pmie command: %s", command)
+        out = subprocess.Popen(
+            [command], stdout=subprocess.PIPE, stderr=subprocess.STDOUT, shell=True
+        )
+        stdout, stderr = out.communicate()
+        self.logger.debug("stop pmie command output:%s", stdout.decode())
+
+        if stdout.decode() != "":
+            # pmie failed. Log it
+            self.logger.error(
+                "not able to stop pmie successfully due to:{}".format(stdout)
+            )
+        else:
+            self.logger.info("pmie stopped successfully")
+
+    def check_connection(self, hosts):
+        """ Check if pmcd is running in given list of hosts"""
+        self.logger.info("checking connection")
+
+        for host in hosts:
+            res = ""
+            try:
+                self.logger.debug("checking connection on host:%s", host)
+                res = pmapi.pmContext(api.PM_CONTEXT_HOST, host + ":44321")
+            except Exception:
+                res = None
+
+            if res is not None:
+                self.logger.info("Port 44321 is open in:{}".format(host))
+                self.hosts.append(
+                    host
+                )  # logging only on those hosts with port 44321 active
+            else:
+                self.logger.error("Port 44321 is not open in:{}".format(host))
+                self.logger.error("Ignoring host:%s", host)
```

Comment:
This shouldn't be repeated twice. Why not promote it to `BaseCollector` (or a new intermediary PCP class) so they can share?